src/requestLib.js

```diff
@@ -0,0 +1,134 @@
+          data: []
+        };
+
+        response.on('data', function (chunk) {
+          responseToSend.data.push(chunk);
+        });
+
+        response.on('end', function () {
+          responseToSend.data = Buffer.concat(responseToSend.data).toString();
+          resolve(responseToSend);
+        });
+      });
+
+      // Log the request that will be initiated on behalf of the client
+      request.on('finish', function () {
+        RdGlobalConfig.ReqLogger.info('Tool Request - Retries Left: ' + retries, clientRequest.method + ' ' + clientRequest.url,
+          false,
+          Object.assign({}, params.furtherRequestOptions, {
+            data: Buffer.concat(params.request.data).toString()
+          }),
+          clientRequest.id);
+      });
+
+      // Capture any error scenarios while making the request on behalf of the client
+      request.on('error', function (err) {
+        reject({
+          message: err,
+          customTopic: 'Tool Request - Retries Left: ' + retries
+        });
+      });
+
+      // Set a hard timeout for the request being initiated.
+      request.setTimeout(constants.CLIENT_REQ_TIMEOUT, function () {
+        request.destroy(constants.REQ_TIMED_OUT);
+      });
+
+      /**
+       * If its the first try of the request, set up all the event listeners to capture/collect
+       * the data being sent by the client.
+       * If its not the first try, then we already have the data to recreate the request
+       * if the previous try fails.
+       */
+      if (retries === constants.MAX_RETRIES) {
+        clientRequest.on('data', function (chunk) {
+          params.request.data.push(chunk);
+          if (!request.write(chunk)) {
+            clientRequest.pause();
+            request.once('drain', function () {
+              clientRequest.resume();
+            });
+          }
+        });
+  
+        clientRequest.on('error', function (err) {
+          request.end();
+          reject({
+            message: err,
+            customTopic: 'Request - Retries Left: ' + retries
+          });
+        });
+  
+        clientRequest.on('end', function () {
+          RdGlobalConfig.ReqLogger.info("Request End", params.request.method + ' ' + params.request.url, false, {
+            data: Buffer.concat(params.request.data).toString()
+          },
+          clientRequest.id);
+          request.end();
+        });
+      } else {
+        request.write(Buffer.concat(params.request.data));
+        request.end();
+      }
+    });
+  },
+
+  /**
+   * Handler for performing request. Includes the retry mechanism when request fails.
+   * @param {{request: Object, furtherRequestOptions: Object}} params 
+   * @param {http.IncomingMessage} clientRequest 
+   * @param {Number} retries 
+   */
+  call: function (params, clientRequest, retries) {
+    retries = (typeof retries === 'number') ? Math.min(constants.MAX_RETRIES, Math.max(retries, 0)) : constants.MAX_RETRIES;
```

Comment:
Configurability for retries?

Reply:
Have added the configurability for the delay after which the retry should be made. That can be set via the CLI args. That should be enough to cater the needs? 1 retry after xyz seconds? Do we need multiple retries?

Reply:
Not sure for the need, may be added if requested from users.

Reply:
Let's see if the requirement comes up and will incorporate that in the next versions?

Reply:
Cool 👍